### ISSUE TITLE
Feature request: Type Parameter Bounds

### ISSUE THREAD

**Issue by tlubz** [ruby/rbs#829]
It's currently difficult to write a well-typed generic container that calls any methods on members of the type parameter. Example:

```ruby
class TestContainer
  attr_reader :obj

  def initialize(obj)
    @obj = obj
  end

  def ==(other)
    @obj == other.obj
  end
end
```

```rbs
class TestContainer[T]
  attr_reader obj: T
  def initialize: (T obj) -> void
  def ==: (untyped other) -> bool
end
```

`steep` (default settings) output:

```
steep check domain/TestContainer.rb 
# Type checking files:

..........................................................F

domain/TestContainer.rb:9:9: [error] Type `T` does not have method `==`
│ Diagnostic ID: Ruby::NoMethod
│
└     @obj == other.obj
           ~~

Detected 1 problem from 1 file
```

Ideally we would be able to assert at least one-sided bounds on the type parameters, for example if we could assert that the parameter inherits from `Numeric`:

```rbs
class TestContainer[T < Numeric]
  attr_reader obj: T
  def initialize: (T obj) -> void
  def ==: (untyped other) -> bool
end
```

**Comment by soutaro** [comment]
Hi @tlubz! 👋 

Yes, we know the requirement of type parameter bounds and have it in our roadmap to RBS 2.0, which will be released with Ruby 3.1 hopefully.
I've started implementation of the feature and will share the progress on this repo soon.

**Comment by pocke** [comment]
This feature has been implemented in https://github.com/ruby/rbs/pull/844. So I'll close this issue.
